### PR TITLE
[cmd/builder] Update builder flags precedence

### LINKED ISSUE #5704: builder flags do not override configuration values
**Describe the bug**
If a value is given in both the input config file and the command flags, the command flag should have precedence (since it is more specific). The value from the input file seems to win.

**Steps to reproduce**
```
$ cat example.conf
dist:
  name: otelcol-dev
  description: "Basic OTel Collector distribution for Developers"
  output_path: ./otelcol-dev

exporters:
  - gomod:
      "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/jaegerexporter
      v0.53.0"
  - import: go.opentelemetry.io/collector/exporter/loggingexporter
    gomod: go.opentelemetry.io/collector v0.53.0

receivers:
  - import: go.opentelemetry.io/collector/receiver/otlpreceiver
    gomod: go.opentelemetry.io/collector v0.53.0

processors:
  - import: go.opentelemetry.io/collector/processor/batchprocessor
    gomod: go.opentelemetry.io/collector v0.53.0
```
```
$ ./builder --name james --output-path /tmp/collector example.conf
2022-07-19T15:45:32.964+1000	INFO	internal/command.go:119	OpenTelemetry Collector Builder	{"version": "dev", "date": "unknown"}
2022-07-19T15:45:32.965+1000	INFO	internal/command.go:138	Using config file	{"path": "example.conf"}
2022-07-19T15:45:32.965+1000	INFO	builder/config.go:99	Using go	{"go-executable": "/opt/homebrew/bin/go"}
2022-07-19T15:45:32.966+1000	INFO	builder/main.go:76	Sources created	{"path": "./otelcol-dev"}
2022-07-19T15:45:33.146+1000	INFO	builder/main.go:108	Getting go modules
2022-07-19T15:45:33.178+1000	INFO	builder/main.go:87	Compiling
2022-07-19T15:45:33.444+1000	INFO	builder/main.go:94	Compiled	{"binary": "./otelcol-dev/otelcol-dev"}
```

**What did you expect to see?**

```
2022-07-19T15:48:13.229+1000	INFO	builder/main.go:76	Sources created	{"path": "/tmp/collector"}
...
2022-07-19T15:48:14.444+1000	INFO	builder/main.go:94	Compiled	{"binary": "/tmp/collector/james"}
```

**What did you see instead?**
see above

**What version did you use?**
Version:  c3a3160ab7fa1ec9d73f1b2ead72127c3a8bf17b

**What config did you use?**
Config:  see above

**Environment**
OS: macos
Compiler(if manually compiled): go 1.18.4

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
This PR updates the builder's config initialization so that flags always take precedence over configuration files or default values.

I opted not to use the collector's resolvers/converter approach as it felt overkill for this command.  This approach takes advantage of the command's existing use of global vars, which isn't the best, but is at least consistent.

**Link to tracking Issue:** 
Fixes #5704 

**Testing:** 
Tested locally

**Documentation:**
I didn't update the README because it feels like flags taking precedence is the norm and the flags are already called out.

### CODE REVIEW

**Comment by TylerHelmuth** [comment]
There is a defect with this solution, closing until I sort it out.

File: cmd/builder/internal/command.go [hunk 1]
```diff
@@ -100,11 +110,44 @@ func initConfig() error {
...
 	}
 
+	cfg.Exporters = cfgFromFile.Exporters
+	cfg.Extensions = cfgFromFile.Extensions
+	cfg.Receivers = cfgFromFile.Receivers
+	cfg.Processors = cfgFromFile.Processors
+	cfg.Replaces = cfgFromFile.Replaces
+	cfg.Excludes = cfgFromFile.Excludes
+
+	if !flags.Changed(SkipComplicationFlag) && cfgFromFile.SkipCompilation {
```
Comment (TylerHelmuth) [inline cmd/builder/internal/command.go:126]: There is some cleanup to do here but wanted to get the basic idea out there:

1. Let cobra set the flags in cfg struct. 
2. Read in the config file
3. Apply all entries to cfg from the config file that cannot be overridden by a flag
4. Check if a flag has not been changed and if the config file included a value. If true, use the config file value. 

We can't read the config file right into cfg or we'll overwrite potentially set flag values.

File: cmd/builder/internal/command.go [hunk 2]
```diff
@@ -24,11 +24,21 @@ import (
 	"github.com/knadh/koanf/providers/file"
 	"github.com/knadh/koanf/providers/posflag"
 	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
 	"go.uber.org/zap"
 
 	"go.opentelemetry.io/collector/cmd/builder/internal/builder"
 )
 
+const SkipComplicationFlag = "skip-compilation"
```
Comment (jpkrohling) [inline cmd/builder/internal/command.go:33]: I personally find it easier to read when this is grouped, like:

```golang

const (
    SkipComplicationFlag = ...
    DistributionNameFlag = ...
)
```

Note that we should skip the compilation, not the complication. We have to remain complicated!

Comment (TylerHelmuth) [reply cmd/builder/internal/command.go:33]: Lol I'll make the change

Comment (TylerHelmuth) [reply cmd/builder/internal/command.go]: Done, although I stand by SkipComplication 😄

**Review by jpkrohling — APPROVED** [review]

File: cmd/builder/internal/command.go [hunk 3]
```diff
@@ -24,11 +24,23 @@ import (
...
 
+const (
+	SkipCompilationFlag            = "skip-compilation"
+	DistributionNameFlag           = "name"
+	DistributionDescriptionFlag    = "description"
+	DistributionVersionFlag        = "version"
+	DistributionOtelColVersionFlag = "otelcol-version"
+	DistributionOutputPathFlag     = "output-path"
+	DistributionGoFlag             = "go"
+	DistributionModuleFlag         = "module"
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply cmd/builder/internal/command.go:41]: No need for them to be public, I'll update.

[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply cmd/builder/internal/command.go]: Done

**Comment by bogdandrutu** [comment]
Please rebase to resolve conflicts.

**Comment by TylerHelmuth** [comment]
> Please rebase to resolve conflicts.

done

**Review by jpkrohling — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**